src/hooks/login/useWalletConnectV2Login.ts

```diff
@@ -0,0 +1,243 @@
+      };
+
+      const loginData = {
+        logoutRoute: logoutRoute,
+        loginType: 'walletconnectv2',
+        callbackRoute: callbackRoute ?? window.location.href
+      };
+
+      if (hasSignature) {
+        dispatch(setWalletConnectLogin(loginData));
+        dispatch(setTokenLoginSignature(signature));
+      } else {
+        dispatch(setWalletConnectLogin(loginData));
+      }
+
+      dispatch(loginAction(loginActionData));
+      optionalRedirect(callbackRoute, onLoginRedirect);
+    } catch (err) {
+      setError(WalletConnectV2Error.invalidAddress);
+      console.error(err);
+    }
+  }
+
+  async function initiateLogin(loginProvider = true) {
+    const shouldGenerateWcUri = loginProvider && !wcUri;
+    if (
+      !walletConnectV2ProjectId ||
+      !walletConnectV2RelayAddress ||
+      (providerRef?.current?.isInitialized?.() && !shouldGenerateWcUri)
+    ) {
+      return;
+    }
+
+    const providerHandlers = {
+      onClientLogin: handleOnLogin,
+      onClientLogout: handleOnLogout
+    };
+
+    const newProvider = new WalletConnectProviderV2(
+      providerHandlers,
+      chainId,
+      walletConnectV2RelayAddress,
+      walletConnectV2ProjectId
+    );
+
+    await newProvider.init();
+    setAccountProvider(newProvider);
+    setWcPairings(newProvider.pairings);
+    providerRef.current = newProvider;
+    if (loginProvider) {
+      generateWcUri();
+    }
+  }
+
+  async function connectExisting(pairing: PairingTypes.Struct | undefined) {
+    if (!walletConnectV2RelayAddress || !walletConnectV2ProjectId) {
+      setError(WalletConnectV2Error.invalidConfig);
+      return;
+    }
+    if (!pairing || !pairing.topic) {
+      return;
+    }
+
+    try {
+      const { approval } = await providerRef.current?.connect({
+        topic: pairing.topic
+      });
+      if (token) {
+        dispatch(setTokenLogin({ loginToken: token }));
+      }
+
+      try {
+        await providerRef.current?.login({ approval });
+      } catch {
+        await initiateLogin();
+      }
+    } catch {
+      setError(WalletConnectV2Error.sessionExpired);
+    } finally {
+      setWcPairings(providerRef.current?.pairings);
+    }
+  }
+
+  async function generateWcUri() {
+    if (!walletConnectV2RelayAddress || !walletConnectV2ProjectId) {
+      setError(WalletConnectV2Error.invalidConfig);
+      return;
+    }
+
+    const { uri, approval } = await providerRef.current?.connect();
+    const walletConnectUri: string | undefined = uri;
+    const hasUri = Boolean(walletConnectUri);
+
+    if (!hasUri) {
+      return;
+    }
+
+    if (!token) {
+      setWcUri(walletConnectUri as string);
+    } else {
+      const wcUriWithToken = `${walletConnectUri}&token=${token}`;
+
+      setWcUri(wcUriWithToken);
+      dispatch(setTokenLogin({ loginToken: token }));
+    }
+
+    try {
+      await providerRef.current?.login({ approval });
+    } catch {
+      await initiateLogin();
```

Comment:
usually, every error should be logged. Please log the error here before calling `initiateLogin()`